ocaml/lambda/translcore.ml

```diff
@@ -1549,16 +1547,16 @@ and transl_scoped_exp ~scopes sort expr =
 
 (* Decides whether a pattern binding should introduce a new scope. *)
 and transl_bound_exp ~scopes ~in_structure pat sort expr loc attrs =
-  let should_introduce_scope =
+  let should_introduce_scope, zero_alloc =
     match expr.exp_desc with
-    | Texp_function _ -> true
-    | _ when in_structure -> true
-    | _ -> false in
+    | Texp_function { zero_alloc } -> true, zero_alloc
+    | _ when in_structure -> true, Default_check
+    | _ -> false, Default_check in
```

Comment:
This is not obviously the same behavior: in `| _ when in_structure -> true` case we could have a non-empty "assume" that would have been propagated. Maybe it was a bug before? Or not feasible? 

Independently, it's a bit strange to use `check_allowed` here in transcore, because it is a parsing concern. We can convert to `Assume_info.t` earlier, in typecore.

Reply:
Well spotted.  To summarize a discussion had in person, I think this is fine, though we can make an additional simplification.

This isn't a bug, because: In the old world, lambda needs to look for attributes in two places to see whether to extend the scope - here we looked for attributes on the let binding and in `transl_function` we looked for attributes on the function.   The effect of the old code here was to add assume info on let bindings into the scope.  The front-end now moves that info into the `Texp_function` (or gives an error if the bound thing is not a function). So it's fine to only look for it in the `Texp_function` case.

But actually, we don't even need to look for it in the `Texp_function` case, because in the new world lambda doesn't need to worry about `zero_alloc` attributes on lets - they are moved to the let-bound function in type checking.  So in  the common case where you have a let-bound function, I was needlessly calculating the `assume_info` and updating the scope twice.  I am pushing a commit that just uses `Assume_info.none` here, because it will be updated in transl_function.